pkg/provisioningv2/rke2/planner/planner.go

```diff
@@ -1248,6 +1238,57 @@ func (p *Planner) ensureRKEStateSecret(controlPlane *rkev1.RKEControlPlane, full
 	}, nil
 }
 
+func getIPAddresses(secrets corecontrollers.SecretCache, machine *capi.Machine) (string, string) {
+	internalIPAddress, internalAddressProvided := machine.Annotations[InternalAddressAnnotation]
+	ipAddress, addressProvided := machine.Annotations[AddressAnnotation]
+
+	// Unmanaged machines won't have a state secret
+	if machine.Annotations["objectset.rio.cattle.io/id"] != "unmanaged-machine" && (!internalAddressProvided || !addressProvided) {
+		secret, err := secrets.Get(machine.Spec.InfrastructureRef.Namespace, name.SafeConcatName(machine.Spec.InfrastructureRef.Name, "machine", "state"))
+		if err == nil && secret != nil && len(secret.Data["extractedConfig"]) != 0 {
+			driverConfig, err := nodeconfig.ExtractConfigJSON(base64.StdEncoding.EncodeToString(secret.Data["extractedConfig"]))
+			if err == nil && len(driverConfig) != 0 {
+				if !addressProvided {
+					ipAddress = convert.ToString(values.GetValueN(driverConfig, "Driver", "IPAddress"))
+				}
+				if !internalAddressProvided {
+					internalIPAddress = convert.ToString(values.GetValueN(driverConfig, "Driver", "PrivateIPAddress"))
+				}
+			}
+		}
+	}
+
+	return internalIPAddress, ipAddress
+}
+
+func appendToSlice(config map[string]interface{}, key string, values ...interface{}) {
+	if len(values) != 0 {
+		switch s := config[key].(type) {
+		case nil:
+			config[key] = values
+		case []interface{}:
+			config[key] = append(s, values...)
+		default:
+			config[key] = append([]interface{}{s}, values...)
+		}
+	}
+}
+
+// getControlPlanPublicIPs returns an []interface{} because the result will be passed to appendToSlice, making it easier to use the ... notation.
+func getControlPlanePublicIPs(secrets corecontrollers.SecretCache, plan *plan.Plan) []interface{} {
+	ips := make([]interface{}, 0)
+	for _, machine := range plan.Machines {
+		if isOnlyWorker(machine) {
+			continue
+		}
+		if _, ipAddress := getIPAddresses(secrets, machine); ipAddress != "" {
+			ips = append(ips, ipAddress)
+		}
+	}
+
+	return ips
```

Comment:
You must sort this slice to ensure that the planner is idempotent, otherwise the hashes in the planner will change.

Reply:
Learned this the hard way last night. I thought I included that change here, but didn't.

Regardless, it's gone now.